openhtf/__init__.py

```diff
@@ -394,6 +394,33 @@ def with_args(self, **kwargs):
     new_info.measurements = [m.with_args(**kwargs) for m in self.measurements]
     return new_info
 
+  def with_plugs(self, **subplugs):
+    """Substitute plugs for placeholders for this phase."""
+    # Make a copy so we can have multiple of the same phase with different plugs
+    # in the same test.
+    new_info = mutablerecords.CopyRecord(self)
+    
+    for substitute_name, substitute_class in subplugs.iteritems():
+      # self.plugs contains PhasePlug so enumerate and unpack
+      for i, original_plug in enumerate(self.plugs):
```

Comment:
It's O(n*m) where n and m are never expected to be more than, say, 50. I'd avoid any performance-based changes for what will never be more than 2500 iterations, if more than 10.

However, if you really cared, The main reason for the performance hit is the O(n) lookup of a plug with the right name. This will speed up the lookups without making the code have randoms `.pop()` calls:

```py
plugs_by_name = {plug.name: plug for plug in self.plugs}
new_plugs = dict(plugs_by_name)

for name, sub_class in subplugs.iteritems():
  original_plug = plugs_by_name.get(name)
  if (original_plug is None
      or not isinstance(original_plug.cls, plugs.PlugPlaceholder)
      or not isinstance(sub_class, original_plug.cls.base_class)):
    raise plugs.InvalidPlugError('...')
  new_plugs[name] = mutablerecords.CopyRecord(original_plug, cls=sub_class)
return mutablerecords.CopyRecord(self, plugs=new_plugs.values())
```

It seems to have also made it less complex, so it has that going for it rather than O(N*M)